Create a login and signup form with validation

xuananh-tran opened the pull request:
## Title

Create Frontend Login And Sign-Up Form With Basic Input Field Validation

# Description

1. Add components to components/ui from shadcn/ui
- Install radix-ui/react-label in server/
- create label/index.tsx
- create input/index.tsx

2. Add login-form.tsx to components/layout:
- Create form with React components
- Only check if all fields (email + password) are filled
- Create buttons for form submission, forgot password, and switch to create account.
- style with shadcn/ui

3. Add signup-form.tsx to components/layout:
- Create form with React components
- Fields: First Name, Last Name, Academic Email, Password, Confirm Password
- Check and show error message for all fields before submission (password and confirm password fields have separate file for validation in utils; password conditions are met when all error messages cleared)
- Create buttons for form submission and switch to log in.
- style with shadcn/ui

4. Add authValidation.ts to lib/utils/
- Check if password length > 8, contains at least 1 lowercase letter, 1 uppercase letter, 1 number, 1 special character
- Check if confirm password is the same as password

5. Create login and signup pages with half-screen form and half-screen image
- Add form components from components/ui
- Upload image to @/public (exported from Figma - the image quality is actually low)
- Style with shadcn/ui

## Type of change

Select option(s) that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] This change requires a documentation update
- [ ] Anything else?

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

Run npm install to get radix-ui/react-label in server/
Probably need to add darkMode: 'class', to tailwind.config.css to enable darkmode in global.css (which I couldn't commit because of eslint!!)

Test:

https://github.com/user-attachments/assets/90a986ef-5e45-40da-87aa-bb545354a9f4



## Additional context

Hours of researching to debug in tears. Sorry for the late PR :(

minhduc29 commented on client/src/app/(routes)/(auth)/auth-form.tsx:
```diff
@@ -0,0 +1,166 @@
+  const { authenFormData } = useUserStore();
+  const [error, setError] = useState<string | null>(null);
+
+  useEffect(() => {
+    const passwordError = checkPasswordStrength(authenFormData.password);
+    const confirmPasswordError = validateConfirmPassword(
+        authenFormData.password,
+        authenFormData.confirmPassword
+    );
+    if (authenFormData.password.length > 0) {
+        setError(passwordError.join('\n'));
+        if (authenFormData.confirmPassword.length > 0 && confirmPasswordError) {
+            setError(confirmPasswordError);
+        } 
+        else {
+            setError(null);
+        }
+    }
+   });
```
What does this part do? What is the rationale for using useEffect here? This means that these validations run every rerender. Why not use onBlur or debounce or another approach?

linhle15-ll replied:
I was thinking of validation on every keystroke, so I useEffect, but you can change it to debouncing to improve UX/performance since debouncing waits until the user stops typing for a few milliseconds before running validation. @xuananh-tran

minhduc29 commented on client/src/app/globals.css:
```diff
@@ -36,6 +36,6 @@ body {
 
   .input-field {
-    @apply flex flex-col gap-2 w-full sm:w-1/2 sm:pr-4 sm:border-r border-gray-300 sm:items-start
+    @apply border rounded-md border-lightGray py-1 px-2 focus:border-primaryOrange focus:outline-none focus:ring-0 focus:ring-primaryOrange;
```
Is this intended to be applied globally or just the input fields in this PR?

minhduc29 replied:
This class is also used in the landing page if I remember correctly.

linhle15-ll replied:
shjet it is :))) consider using another name, like add form-, or auth-, I will choose one styling for input field later @xuananh-tran

minhduc29 commented on client/src/lib/utils/authValidation.ts:
```diff
@@ -1,6 +1,6 @@
-//TODO: validate academic email
+//validate academic email
```
Should academic email validation be in the frontend or backend?

xuananh-tran commented:
@linhle15-ll  @minhduc29 Can you help me review the changes ><?
I made a few changes to this branch.
1. Rewrite user.store to handle access token. I'm focusing only on the auth parts, so I've included only the methods related to auth. Updates to all other fields (bio, profile image) should be implemented later on when we want to update the user profile. Also, the current user store persists the auth form fields even after refresh or successful signup/signin. This is insecure, so I made some updates.
2. Updates to auth-form:
- Use modified user store
- Update useEffect logic: the current logic incorrectly handles the error message (no error message for sign up). I also updated the dependencies array to prevent validation on every rerender.
- Add handle submit function to connect to backend. Currently, after successful sign-up/sign-in, I'm redirecting to the home page ('/') (not shown in tests below), but specific redirection needs to be handled later on. The home page is not running properly, and successful registration with email verification should redirect to log in page.
- Minor updates to naming convention on both the auth backend and frontend.

Note:
I store the access token in zustand memory and not local storage because I thought local storage might not be secure. I'm not sure if using zustand store is better. Let me know whether this is a good approach.

Tests:
Sign in:

https://github.com/user-attachments/assets/7dec8899-a2a4-43ea-b866-aab938e2d44c


Backend properly connected:
![image](https://github.com/user-attachments/assets/66ec6df6-33e6-48ec-b979-207ebeb8e519)
Access token is returned in response:
![image](https://github.com/user-attachments/assets/039f9561-172b-4b32-bce7-ce4ad09d494c)
Cookies with refreshtoken:
![image](https://github.com/user-attachments/assets/ce9d431f-60ac-4c2f-b642-b2a4087eb1ee)
Access token stored in zustand store (checked using Redux Devtools Extension). It is the same access token in response tab.
![image](https://github.com/user-attachments/assets/bb10e08d-9f07-4b10-85f9-9f77a22c48c9)
Access token is not stored in local storage:
![image](https://github.com/user-attachments/assets/d3676c9a-ffe6-430b-a993-9236c717e0d4)


Sign up:
Current sign up flow does not include generation of any tokens.

https://github.com/user-attachments/assets/60dd8980-c0b8-4abb-b08a-86333c5419d3



Backend connected:
![image](https://github.com/user-attachments/assets/d9a06049-b825-406c-91f8-e9cfebd7e131)
![image](https://github.com/user-attachments/assets/348cfc9a-f636-4a88-a329-fcf1a35618fd)
![image](https://github.com/user-attachments/assets/8a5a6775-1cec-4f27-9ba2-9a5e1c671890)


Not shown in tests, but links redirecting to sign up and sign in are also working properly

minhduc29 commented on client/src/app/(routes)/(auth)/auth-form.tsx:
```diff
@@ -0,0 +1,227 @@
+        ? 'http://localhost:5001/api/auth/signup'
+        : 'http://localhost:5001/api/auth/signin';
+
+      const res = await fetch(url, {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify(formData),
+      });
+
+      const data = await res.json();
```
We have Axios set up for the purpose of sending requests from the frontend to the backend. Take a look at the `services/` folder to see what it does and how to implement and use a service function.

minhduc29 commented on client/src/lib/stores/user.store.ts:
```diff
@@ -0,0 +1,78 @@
+
+export const useUserStore = create<UserStore>()(
+  devtools(
+    subscribeWithSelector(
```
I haven't really looked into the details of this file, but here are some things that instantly came to my mind that you may want to address:
- Is it necessary to use zustand/store for all of these fields? (Are they supposed to be shared globally?)
- What is the use case of subscribeWithSelector here? When and where is it used?
- Why should there be initial form values?
- Why were the fields listed manually without the use of anything from `types/`? Is the user type not reusable?
- Which ones of these fields need to be persisted?

minhduc29 commented on server/src/models/user.model.ts:
```diff
@@ -23,7 +23,7 @@ const userSchema = new Schema<User>(
       trim: true,
       lowercase: true,
-      match: /^[^\s@]+@[^\s@]+\.(edu)$/,
+      // match: /^[^\s@]+@[^\s@]+\.(edu)$/,
```
If this is not necessary, just get rid of it instead of commenting it out.

minhduc29 commented on server/src/routes/auth.routes.ts:
```diff
@@ -4,8 +4,8 @@ import authController from '../controllers/auth.controller';
+router.post('/signin', authController.handleSignIn);
 router.post('/refresh', authController.handleRefreshToken);
-router.post('/logout', authController.handleLogOut);
+router.post('/signout', authController.handleSignOut);
```
I like these routes' paths.

xuananh-tran commented:
@linhle15-ll @minhduc29 I made some changes:
- Rewrote `user.store`. I think it's right to consider what we want to store globally. Because the states for auth form are only necessary for authentication, I manage them locally in `auth-form`. The flow is: manage auth form states locally --> verification --> get verified returned `user` and `accessToken` from backend --> store and manage them globally in `user-store`. I think this better separates concerns and keeps the global `user-store` clean. @linhle15-ll Please let me know if you still want to handle auth states globally in store. I guess I'll try to come up with a better flow.
- Current auth flow: sign up --> successful --> sign in --> no access token returned and nothing stored in user-store.
Sign-in --> access token returned with user information --> stored in local storage.
- Notes on `user-store`: I keep `devtools` because it doesn't hurt to keep it, `subscribeWithSelector` because this is helpful when handling tokens later on (in another PR :( ), and `persist` with `partialize` because we need to keep access token and other user information (currently id and email) in localStorage.
- I addressed comments about `globals.css`, user types and models (because we can verify email with the API). Please keep in mind (just in case you do tests later on) that we are limited to having 5 API calls a day only. Having email verification in model (with .edu, for example, even though we agreed that .edu check is unreliable) reduces the burden to make API calls, but I deleted that.
- I changed auth-form for better UX: use onBlur for password-related errors, add show password buttons, and group all error messages into one message. I also put error messages for password fields right below the corresponding input fields.
- Used axios instead of fetch.
- Minor naming changes in backend for consistency.
Notes: I should probably handle loading state later on.

Tests:
Sign-in (redirects to home):
Improved UX (I hope):

https://github.com/user-attachments/assets/55292ff1-6130-4a47-ac9a-382875a96d69


Backend properly connected. Refresh token in cookie.
![image](https://github.com/user-attachments/assets/84cc277f-b62a-4853-b860-9d8daf4530fa)
![image](https://github.com/user-attachments/assets/fe86720f-b1e2-4833-a609-f4a1c29e9d38)
Access token, correct user id and email stored in localStorage. All survive refreshes. Everything else is reset with refreshes.
![image](https://github.com/user-attachments/assets/66bbee08-097d-44c7-b598-880ffd165fb1)

Sign-up (redirects to sign-in; after signing in everything is tested the same as above):
Improved UX (I hope):

https://github.com/user-attachments/assets/25367ad1-8ec3-4a7a-945b-191cccf79f7d


Backend properly connected. No tokens returned. No cookie. Nothing in localStorage.

minhduc29 commented:
LGTM! There can be some improvements (ask me later after check-in or something), but it's generally good for now!

xuananh-tran commented:
@linhle15-ll  @minhduc29 I changed `api.ts` to handle access token. The function looks a bit messy because we are not directly storing accessToken in localStorage but through zustand store. I will test this function later on when handling both access and refresh tokens in frontend.

minhduc29 commented on client/src/app/(routes)/(auth)/auth-form.tsx:
```diff
@@ -0,0 +1,244 @@
+      return;
+    }
+
+    try {
+      const url = isSignUp ? 'auth/signup' : 'auth/signin';
+      const res = await api.post(url, authForm);
+
+      const data = res.data;
+      if (!isSignUp) {
+        setAccessToken(data.accessToken);
+        setUser(data.user);
+
+        router.push('/home');
+      } else {
+        router.push('/sign-in');
+      }
+    } catch (error: unknown) {
+      const axiosError = error as AxiosError<{ error: string }>;
+
+      const message =
+        axiosError.response?.data?.error ||
+        'An error occurred. Please try again.';
+
+      setError(message);
+    }
```
This is the part that I said to ask me later after check-in. This part is functional I suppose, so you don't need to change anything here, but they should have been declared separately as service functions in `/services`. I might just change this part manually if we have a PR for cleaning up the whole repo since there are some other changes I want to make as well. Good job on the AxiosError part tho! I didn't know you could do it like that.
